### PR TITLE
WIP: What's New in TM:PE 11.6.5.2

### PR DESCRIPTION
**Pull request by aubergine10**
- **[Meta] This release improves Lane Routing (UI, trams, trains, etc.), Parking AI (electric cars), and icon themes.**
- **[Meta] TM:PE data format change: Old cities will still load (data converted automatically), but after saving you'll need TM:PE 11.6.5.2 (or newer) for those saves.**
- [New] Directional lane connectors #1492 #1214 #1211 (kianzarrin, Chamëleon)
- [New] Lane routing diagnostic tool for developers #1549 #1537 #1534 (kianzarrin)
- [New] Chinese road sign theme #1523 (kvakvs, lokpro, TianQiBuTian)
- [Mod] Incompatible: Two more CSM mods #1481 (StayHungryStayFoolish, aubergine18)
- [Mod] Incompatible: 交通管理器 汉化版TMPE_V11_STABLE #1481 (StayHungryStayFoolish, aubergine18)
- [Mod] Incompatible: Service Vehicles Manager 2.0 clone #1481 (StayHungryStayFoolish, aubergine18)
- [Fixed] Speed limits not working on station tracks #1467 #1337 (Peacemak3r, FemboyPyro, aubergine18, krzychu124)
- [Fixed] Patching `Reversible Tram AI` fails if mod subscribed but disabled #1471 #1470 (kianzarrin)
- [Fixed] Lane connectors buggy on bidirectional tracks #1211 #784 (kianzarrin)
- [Fixed] Zig-zagging routes on bidirectional tracks #1487 #1486 (kianzarrin)
- [Fixed] Excessive track connections at junctions #1483 #1475 (kianzarrin)
- [Fixed] Main menu button breaks on hot-reload #762 #1492 (kianzarrin)
- [Fixed] Asset Editor crash when saving asset with no segments #1505 #1313 (krzychu124, kianzarrin)
- [Fixed] Toggling certain mod options causes blank menu buttons #1516 #1514 (krzychu124)
- [Fixed] Trams unable to turn due to lane arrows #1501 #1053 (WildQuoll, RLemonache, kianzarrin)
- [Fixed] Parking AI: Residents ignoring `Electric Cars` district policy #1543 #1527 (krzychu124)
- [Fixed] Parking AI: Drivers leaving city try to park at outside connection #1543 #1527 (krzychu124)
- [Fixed] Parking AI: Non-dummy tourists not using electric cars #1543 #1527 (krzychu124)
- [Fixed] Parking AI: Residents moving into city never have electric cars #1543 #1527 (krzychu124)
- [Fixed] API: `ManagersFactory` not working #1548 (kianzarrin)
- [Updated] 419 localisations updated #1523 (Dominik Kawula, shg166, John Lok Ho, Skazov, kvakvs, TwotoolusFLY_LSh.st, DucAnhLam, Márcio Saeger, Sanam Rao, Naruebet Chaipin, Nguyễn Tài Đức)
- [Updated] Multiple icons updated, fallback theme, etc. #1523 (kvakvs)
- [Updated] Separated road and tracked vehicle routing in Routing Manager #1501 (kianzarrin)
- [Updated] Trains prefer to stay in lane at intersection #1483 #1475 (kianzarrin)
- [Updated] Mod options: Drop-down lists converted to custom components #1512 #1510 #1468 (kianzarrin)
- [Updated] Improve lane connector UI #1491 #1485 (kianzarrin)
- [Updated] Routing Manager code clean-up #1493 #1477 (kianzarrin)
- [Updated] Replace obsolete `GetSortedLanes` with extension #1522 #1521 #1496 #1490 (aubergine18)
- [Updated] Replace obsolete `IsStartNode` with extension #1506 (aubergine18)
- [Updated] Update `Microsoft.Unity.Analyzers` to version `1.13.0` #1513 (aubergine18)
- [Updated] Add UI helpers for mod option sliders #1531 #1512 #1510 (kianzarrin)
- [Removed] Flawed TTL API removed to prevent adoption #1495 (Elesbaan70)

### CODE REVIEW

**Comment by aubergine10** [comment]
Looks like 11.6.5.2 is about ready. Will need to send another commit to this PR to update changelogs.

If anyone has any stuff that needs going in 11.6.5.2 lmk asap.

**Comment by aubergine10** [comment]
Is there no way we can fix that issue? Both the global config and the mod options now have version numbers so we can potentially translate back and fourth (might require a cherrypicked update to STABLE)?

Otherwise how are we going to get 11.6.5.2 out to test?

**Comment by aubergine10** [comment]
Which aspects of the serialized data changed? Is it something that failed load will only break one tool, or does it break everything?

**Comment by krzychu124** [comment]
> Which aspects of the serialized data changed?

https://github.com/CitiesSkylinesMods/TMPE/pull/1211/files#diff-f665322d253fe3555fd1877d682fa1e6e219ae4e744486c93c98286d0d5d4f4e

>  Is it something that failed load will only break one tool, or does it break everything?

Everything. Binary serialization is "all or nothing" if you don't explicitly add methods to handle changed things in advance. 
I'll say that once again because I'm aware it might not be obvious where is the problem. `Master` or `11.6.5.2` **can read older data** _(lane connection settings, arrows, speed limits etc.)_ but saved games created with that version **cannot be read** by the **older version** of the mod (eg. current workshop Stable or anything older) because they don't expect change and don't have handlers to e.g. ignore or try transform unknown data. 
Since we hold everything in one Configuration class and one data container then it will be complete wipe out of data from the user perspective.

I think that reverting names in that class to previous model would fix the problem but still requires implementing additional handling to read data from current `master` into original old version scheme. In any way it is a mess we need to fix as soon as possible changing serialization method and also change in the way we store things (data container per manager, not all in one)

<details><summary>Error stacktrace</summary>
<p>

```cs
Info 269.5560148: Loading Data from New Load Routine! Length=8943
Error 269.6905090: Error deserializing data: System.Runtime.Serialization.SerializationException: Field "sourceLaneId" not found in class TrafficManager.Configuration+LaneConnection
  at System.Runtime.Serialization.Formatters.Binary.ObjectReader.ReadTypeMetadata (System.IO.BinaryReader reader, Boolean isRuntimeObject, Boolean hasTypeInfo) [0x00000] in <filename unknown>:0 
  at System.Runtime.Serialization.Formatters.Binary.ObjectReader.ReadObjectInstance (System.IO.BinaryReader reader, Boolean isRuntimeObject, Boolean hasTypeInfo, System.Int64& objectId, System.Object& value, System.Runtime.Serialization.SerializationInfo& info) [0x00000] in <filename unknown>:0 
  at System.Runtime.Serialization.Formatters.Binary.ObjectReader.ReadObject (BinaryElement element, System.IO.BinaryReader reader, System.Int64& objectId, System.Object& value, System.Runtime.Serialization.SerializationInfo& info) [0x00000] in <filename unknown>:0 
  at System.Runtime.Serialization.Formatters.Binary.ObjectReader.ReadNextObject (System.IO.BinaryReader reader) [0x00000] in <filename unknown>:0 
  at System.Runtime.Serialization.Formatters.Binary.ObjectReader.ReadObjectGraph (BinaryElement elem, System.IO.BinaryReader reader, Boolean readHeaders, System.Object& result, System.Runtime.Remoting.Messaging.Header[]& headers) [0x00000] in <filename unknown>:0 
  at System.Runtime.Serialization.Formatters.Binary.BinaryFormatter.NoCheckDeserialize (System.IO.Stream serializationStream, System.Runtime.Remoting.Messaging.HeaderHandler handler) [0x00000] in <filename unknown>:0 
  at System.Runtime.Serialization.Formatters.Binary.BinaryFormatter.Deserialize (System.IO.Stream serializationStream) [0x00000] in <filename unknown>:0 
  at TrafficManager.Lifecycle.SerializableDataExtension.DeserializeData (System.Byte[] data) [0x00000] in <filename unknown>:0 
   at CSUtil.Commons.Log.LogToFile(System.String log, LogLevel level)
   at CSUtil.Commons.Log.Error(System.String s)
   at TrafficManager.Lifecycle.SerializableDataExtension.DeserializeData(System.Byte[] data)
   at TrafficManager.Lifecycle.SerializableDataExtension.Load()
   at TrafficManager.Lifecycle.SerializableDataExtension.OnLoadData()
   at SerializableDataWrapper.OnLoadData()
   at SimulationManager.LateUpdateData(UpdateMode mode, Single minProgress, Single maxProgress)
   at SimulationManager.Managers_LateUpdateData(UpdateMode mode, Single minProgress, Single maxProgress)
   at LoadingManager+<LoadSimulationData>c__IteratorB.MoveNext()
   at AsyncTask.Execute()
   at SimulationManager.SimulationStep()
   at SimulationManager.SimulationThread()

Info 269.7049555:   at System.Runtime.Serialization.Formatters.Binary.ObjectReader.ReadTypeMetadata (System.IO.BinaryReader reader, Boolean isRuntimeObject, Boolean hasTypeInfo) [0x00000] in <filename unknown>:0 
  at System.Runtime.Serialization.Formatters.Binary.ObjectReader.ReadObjectInstance (System.IO.BinaryReader reader, Boolean isRuntimeObject, Boolean hasTypeInfo, System.Int64& objectId, System.Object& value, System.Runtime.Serialization.SerializationInfo& info) [0x00000] in <filename unknown>:0 
  at System.Runtime.Serialization.Formatters.Binary.ObjectReader.ReadObject (BinaryElement element, System.IO.BinaryReader reader, System.Int64& objectId, System.Object& value, System.Runtime.Serialization.SerializationInfo& info) [0x00000] in <filename unknown>:0 
  at System.Runtime.Serialization.Formatters.Binary.ObjectReader.ReadNextObject (System.IO.BinaryReader reader) [0x00000] in <filename unknown>:0 
  at System.Runtime.Serialization.Formatters.Binary.ObjectReader.ReadObjectGraph (BinaryElement elem, System.IO.BinaryReader reader, Boolean readHeaders, System.Object& result, System.Runtime.Remoting.Messaging.Header[]& headers) [0x00000] in <filename unknown>:0 
  at System.Runtime.Serialization.Formatters.Binary.BinaryFormatter.NoCheckDeserialize (System.IO.Stream serializationStream, System.Runtime.Remoting.Messaging.HeaderHandler handler) [0x00000] in <filename unknown>:0 
  at System.Runtime.Serialization.Formatters.Binary.BinaryFormatter.Deserialize (System.IO.Stream serializationStream) [0x00000] in <filename unknown>:0 
  at TrafficManager.Lifecycle.SerializableDataExtension.DeserializeData (System.Byte[] data) [0x00000] in <filename unknown>:0 
Error 269.7069267: OnLoadData: Error while deserializing data: System.ApplicationException: An error occurred while loading
  at TrafficManager.Lifecycle.SerializableDataExtension.DeserializeData (System.Byte[] data) [0x00000] in <filename unknown>:0 
  at TrafficManager.Lifecycle.SerializableDataExtension.Load () [0x00000] in <filename unknown>:0 
   at CSUtil.Commons.Log.LogToFile(System.String log, LogLevel level)
   at CSUtil.Commons.Log.Error(System.String s)
   at TrafficManager.Lifecycle.SerializableDataExtension.Load()
   at TrafficManager.Lifecycle.SerializableDataExtension.OnLoadData()
   at SerializableDataWrapper.OnLoadData()
   at SimulationManager.LateUpdateData(UpdateMode mode, Single minProgress, Single maxProgress)
   at SimulationManager.Managers_LateUpdateData(UpdateMode mode, Single minProgress, Single maxProgress)
   at LoadingManager+<LoadSimulationData>c__IteratorB.MoveNext()
   at AsyncTask.Execute()
   at SimulationManager.SimulationStep()
   at SimulationManager.SimulationThread()
```

</p></details>

**Comment by Elesbaan70** [comment]
> - revert the change - will break savegames already created with `master` branch if we do not add code which could optionally try to read changed things, so.... not the best idea

But how bad an idea is that, really? I mean, it's not ideal, but I kind-of expect things like that to happen when I'm pulling source and building it myself.

Note that, assuming my project is ready in time, the new lane connector code's serialization will still have to be reverted to the old way, then redone in the new framework and new backward-compatibility standards.

I dislike the idea of reverting it this late, but if the compatibility issue is a showstopper, then that might be the only viable option, unless you're willing to delay the release a bit.

> - wait for @Elesbaan70 to implement serialization thingy - still may not fix the problem from the previous point

Unfortunately it's not the best time for my work to become time-sensitive. With Pascha coming up, other obligations will take priority for the next two weeks. It might still work out, but I wouldn't want to overcommit.

And no, it won't fix the current breaking change in `master`. We're stuck with that.

> - ignore it, release to TEST then as Stable - in case of any issues with loading data on current version we won't be able return to any previous(e.g.: via "revert to this version" using Steam workshop).

*That* would not be the best idea. 😬

**Comment by krzychu124** [comment]
Yeah, I think we need to revert the change since it is not impossible and then (after serialization thingy will be finished) we will have more ways to save/load data without breaking other not related data.

**Comment by aubergine10** [comment]
How many TEST users are going to try moving to STABLE in the time it takes for us to decide that current TEST is reliable enough to move to STABLE?

**Comment by Elesbaan70** [comment]
> How many TEST users are going to try moving to STABLE in the time it takes for us to decide that TEST is reliable enough to move to STABLE?

If a serious issue is discovered, maybe all of them. 

We *did* rewrite the lane connector.....

**Comment by krzychu124** [comment]
> How many TEST users are going to try moving to STABLE in the time it takes for us to decide that TEST is reliable enough to move to STABLE?

Looking how many PRs @kianzarrin has still in progress, probably few weeks and 11.6.5.2 won't be the last TEST version...

**Comment by kianzarrin** [comment]
I have 3 and 2 of them are debug only

**Comment by aubergine10** [comment]
@kianzarrin do any of your PRs need to go in 11.6.5.2?

**Comment by Elesbaan70** [comment]
Also, an unexpected traffic light API dependency could force people to revert to stable until we either revert that change or work with the mod developer(s) to resolve the issue.

**Comment by aubergine10** [comment]
> Also, an unexpected traffic light API dependency could force people to revert to stable until we either revert that change or work with the mod developer(s) to resolve the issue.

Is that in 11.6.5.2? If so, what mods does it affect?

**Comment by Elesbaan70** [comment]
> Is that in 11.6.5.2? If so, what mods does it affect?

It has been merged into master. Is that your release process, that it will be in 11.6.5.2?

It is not expected to affect any mod, because the belief is that no one uses the traffic light API. There's really no way to verify this belief other than to put it out there and see what happens.

**Comment by aubergine10** [comment]
Ah, sorry, the API removal - yeah, I don't think that will cause any issues as it was a garbled mess anyway.

**Comment by kianzarrin** [comment]
I have only one PR which is a complicated bug fix. It might cause regression. I think it should go to test but not stable

File: CHANGELOG.md [hunk 1]
```diff
@@ -24,11 +24,37 @@ This changelog includes all versions and major variants of the mod going all the
...
->  
-> Date format: dd/mm/yyyy
 
 </details>
 
+> Date format: dd/mm/yyyy
+
+#### TM:PE V11.6.5.2 TEST, 18/04/2022
+
+- [Meta] This release improves lane routing, and fixes some bugs.
```
Comment (kianzarrin) [inline CHANGELOG.md:34]: also lane connector UI/back-end was improved a little bit. Does not count as lane routing from user point of view?

Comment (aubergine10) [reply CHANGELOG.md]: yup, I tweaked the text to include note on UI and the other stuff is listed in more detail elsewhere in the list

**Comment by krzychu124** [comment]
@aubergine10 has anything left for updating here before merge?

**Review by krzychu124 — APPROVED** [review]

**Comment by aubergine10** [comment]
@krzychu124 Just this small PR from Kian related to the routing diagnostic tool: https://github.com/CitiesSkylinesMods/TMPE/pull/1549

**Comment by aubergine10** [comment]
Also https://github.com/CitiesSkylinesMods/TMPE/pull/1523 is now merged - is that going in 11.6.5.2? If so I'll update changelogs.

**Comment by aubergine10** [comment]
Is Vietnamese new language in 11.6.5.2 ? - if so I should probably mention that in changelogs

**Comment by kvakvs** [comment]
> Is Vietnamese new language in 11.6.5.2 ? - if so I should probably mention that in changelogs

I approved the strings but they aren't in the translation files yet. Some manual work is required to add a column to export files, so i decided to not pull that stunt in the Chinese branch.

**Review by kvakvs — APPROVED** [review]

**Review by kianzarrin — APPROVED** [review]

**Comment by krzychu124** [comment]
@aubergine10 ready? ready? 😄

**Comment by aubergine10** [comment]
merged :)

**Pull request merged**